### PR TITLE
Update Quest tooling and throughput spreadsheet workflow

### PR DESCRIPTION
**Pull request by KjellKod**
## Summary
Expand the branch's Quest support tooling and throughput spreadsheet workflow so Quest can be installed and validated from the repo, and the spreadsheet analysis matches the Jira/GitHub process now being used.
- Adds the Quest runtime, manifest/install helpers, validation scripts, review intelligence, and related tests introduced on this branch.
- Adds the engineering throughput spreadsheet skill, year-scoped Jira individual metrics support, and follow-up skill guidance for repo confirmation, Jira/GitHub summary tabs, adjusted GitHub process metrics, and phased improvement goals.
- Pins \\`pathspec\\` to keep \\`black==26.3.1\\` installable.

> [!NOTE]
> Draft review feedback on the Quest runtime portion is being handed to the Quest maintainers separately; this PR keeps the branch scope visible while that follow-up happens.

## Changes
- **Quest tooling**:
  - Add \\`.quest-manifest\\`, \\`scripts/quest_installer.sh\\`, \\`scripts/quest_preflight.sh\\`, \\`scripts/quest_complete.py\\`, \\`scripts/quest_startup_branch.py\\`, and the \\`scripts/quest_runtime/\\` package for runtime, review-cycle, and artifact handling.
  - Add Quest validation, allowlist, review-intelligence, celebration, and checks entrypoints under \\`scripts/quest_*\\` and refresh related skill docs in \\`.agents/skills/\\`, \\`.skills/quest/\\`, and \\`.skills/review-decisions/SKILL.md\\`.
  - Update Quest guidance in \\`.ai/quest.md\\`, \\`.quest-checksums\\`, \\`.quest-version\\`, and related repo metadata.
- **Throughput workflow**:
  - Add \\`.skills/engineering-throughput-spreadsheet/SKILL.md\\` and catalog it in \\`.skills/SKILLS.md\\`.
  - Extend the throughput spreadsheet guidance to require explicit repo confirmation from \\`.env\\`, add \\`Jira Summary\\` and \\`GitHub No Approval\\`, preserve raw vs process-eligible GitHub metrics, and capture phased improvement goals / EM scorecards.
  - Add \\`--year\\` support in \\`jira_metrics/individual.py\\` with coverage in \\`jira_metrics/tests/test_individual.py\\`.
- **Validation and dependencies**:
  - Add tests in \\`tests/unit/test_allowlist_matcher.py\\`, \\`tests/unit/test_review_intelligence.py\\`, \\`tests/unit/test_codex_skill_wrappers.py\\`, and \\`tests/integration/test-enforce-allowlist.sh\\`.
  - Pin \\`pathspec\\` in \\`requirements.txt\\` so \\`black==26.3.1\\` installs cleanly.

## Validation
- [ ] Run \\`bash scripts/quest_validate-manifest.sh\\` and expect \\`All Quest files are listed in .quest-manifest\\` plus a clean stale-entry check.
- [ ] Run \\`python3 -m unittest jira_metrics.tests.test_individual\\` and expect the year-filter coverage to pass.
- [ ] Run \\`python3 -m pytest tests/unit/test_allowlist_matcher.py tests/unit/test_review_intelligence.py tests/unit/test_codex_skill_wrappers.py\\` and expect the Quest support unit tests to pass.
Watch for: draft review feedback on the Quest runtime portion is being tracked separately from this PR.

## Notes
- The branch already contained \\`Add throughput spreadsheet skill and update Quest tooling\\` and \\`Fix pathspec pin for Black 26\\`; this PR also includes the follow-up skill refinements made after the spreadsheet analysis work.

<pre>
     ▐▛███▜▌
    ▝▜█████▛▘
      ▘▘ ▝▝
Quest/Co-Authored by
Co-Authored-By: Codex <noreply@openai.com>
in collaboration with KjellKod <kjell.hedstrom@gmail.com>
</pre>

---
## Summary by cubic
Adds a generic engineering throughput build flow (Jira + GitHub) with bounded Google Sheets payloads and a required agent-authored `Recommendations` step, and hardens the Quest plan‑review workflow with scratch artifacts, atomic publish, and stricter validation. Also removes real org identifiers from tests and adds a guard to prevent future leakage.

- **New Features**
  - Engineering throughput: runtime config and date windows; exclusion‑aware Jira/GitHub summaries; bounded tabs (“Jira Summary”, “GitHub Summary”, “GitHub No Approval”, “GitHub Repos”, “GitHub Authors”, “GitHub Flags”, “Recommendations”); structured `recommendation_signals.json` and mandatory agent‑authored `Recommendations` (human‑only runs stop at local artifacts); `scripts/engineering_throughput_build.py` and `scripts/engineering_throughput_show_config.py`; committed modules under `engineering_throughput/`, `git_metrics/`, and `jira_metrics/`; unit/integration tests and tracked fixtures.
  - Quest: `.next` scratch artifacts and atomic publish; phase‑aware backlogs with strict validation; runner telemetry via `QUEST_RUNNER_TELEMETRY_LOG`; installer cleans up removed managed files and stale checksums; checks CLI runs allowlist + runtime shell tests; expanded manifest; `.gitignore` adds `.worktrees/`, `.ws/`, `.codex/environments/`.

- **Bug Fixes**
  - Throughput: reject future month‑end `date_end`; treat malformed/non‑string dates as errors; split GitHub searches into month windows to avoid the 1,000‑result GraphQL cap; clarify optional `--team-config`; document the agent‑authored recommendations step and Sheets write requirement; remove raw identifiers from tests and add a deny‑list leakage guard.
  - Allowlist: normalize repo‑relative paths; block absolute‑path `find`/`rg`; integration tests cover allowed/blocked and role‑scoped writes.
  - Backlog/state: appends are idempotent; retagging won’t persist on append failure; recheck expected phase under a lock; `quest_complete.py` reports invalid `--date` cleanly.
  - CI/Lint/Docs: add a `pylint` init‑hook for repo‑root imports; docs clarify scratch→publish and runtime dispatch; coding rules consolidated; throughput skill adds Jira/GitHub summary tabs, process‑adjusted metrics, repo confirmation, and the agent‑authored `Recommendations` contract.

<sup>Written for commit 00f819601ffcca6e89073a2b38bf5abe2fbc130e. Summary will update on new commits.</sup>

### CODE REVIEW

File: AGENTS.md.quest_updated [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+| Topic | Location |
+|-------|----------|
+| Quest orchestration | `.ai/quest.md` |
+| SKILLs directory  | `.skills/SKILLS.md` |
+| Quest setup guide | `docs/guides/quest_setup.md` |
+| Architecture | `docs/architecture/` (if present) |
+
+## Quest Execution Discipline
+- For `$quest`, follow the full gate sequence: routing -> plan -> dual plan review -> arbiter -> walkthrough -> explicit approval -> build -> dual code review -> fixes.
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply AGENTS.md.quest_updated]: this file will be removed.

File: .skills/pr-assistant/SKILL.md [hunk 2]
```diff
@@ -163,15 +197,36 @@ Append this block at the end of the PR body:
...
+- If the work was done through Quest, check quest artifacts first to identify participating models.
+- If generating a squash commit for a PR, you may also check PR comments for clearly attributable model participation.
+- Do not require PR metadata in non-PR contexts. If attribution remains uncertain after available checks, omit the unverified model.
+- Use the specific model label (e.g., "Claude Opus 4.6", "Codex mini") when known from the session or quest artifacts.
+- Known model email mappings:
+  - Claude → `noreply@anthropic.com`
+  - Codex / OpenAI → `noreply@openai.com`
+  - OpenCode → `noreply@opencode.ai`
+- Resolve `<human author identity>` from local git config when available:
+  - Prefer `git config user.name` + `git config user.email` and format as `Name <email>`
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .skills/pr-assistant/SKILL.md]: agreed

File: scripts/quest_review_intelligence.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+            raise ValueError("backlog must be a JSON object")
+
+        existing_defer_ids = {
+            str(item.get("finding_id") or "")
+            for item in payload.get("items", [])
+            if isinstance(item, dict) and item.get("decision") == "defer"
+        }
+
+        retagged = retag_backlog_at_cap(payload)
+        _write_json(backlog_path, retagged)
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_review_intelligence.py]: I agree with cubic on that one.

  **Current state:**

  - In quest_review_intelligence.py:211
  - And in scripts/quest_review_intelligence.py:211

  the code still does this order:

  1. retag backlog in memory
  2. write review_backlog.json
  3. then append deferred records to .quest/backlog/deferred_findings.jsonl

  **Why that is a real bug:**

  - if the append step fails, the backlog now says “defer” but the deferred-history reservoir never got the record
  - that loses carry-over lineage, which is exactly the data this path is supposed to preserve
  - downstream features that rely on deferred-history can silently miss those items

  **My nuance on it:**

  - the comment is directionally correct, but a simple reorder is not a full transaction
  - if we only switch to “append first, then write backlog”, we avoid lost history on append failure, but a later backlog-write failure can leave appended records behind and cause duplicates on retry
  - so the proper fix is either:
      1. make deferred append idempotent by key, then append first and write backlog second, or
      2. add a small journal/temp-file commit pattern across the two outputs

  **So my honest take:**

  - yes, this is still valid
  - yes, we should fix it in Quest source
  - no, I would not call a pure reorder the complete solution
  
  
  **Quest will update this and fix it externally and we will update after that has happened**

File: scripts/quest_allowlist_matcher.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+def shell_tokens(command: str) -> list[str] | None:
+    try:
+        return shlex.split(command, posix=True)
+    except ValueError:
+        return None
+
+
+def contains_blocked_find_action(command_tokens: list[str]) -> bool:
+    if not command_tokens or command_tokens[0] != "find":
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_allowlist_matcher.py]: • On r3132391614: I agree with cubic there too. The issue is real because the matcher treats find as blocked only when the first token is literally "find". /usr/bin/find or /bin/find bypasses that branch, so non-
  exact allowlist entries can still carry -exec or -delete. 
  
  The right fix is to normalize the executable to its basename before applying the find/rg safety checks, and cover that with tests. I’m folding that into  the source repo while I’m fixing the deferred-write bug.
  
  **Quest will update this and fix it externally and we will update after that has happened**

File: scripts/quest_allowlist_matcher.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        return False
+    return any(token in BLOCKED_FIND_ACTIONS for token in command_tokens[1:])
+
+
+def contains_blocked_rg_flag(command_tokens: list[str]) -> bool:
+    if not command_tokens or command_tokens[0] != "rg":
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_allowlist_matcher.py]: will be fixed with https://github.com/KjellKod/metrics-and-insights/pull/113#discussion_r3132905083

File: AGENTS.md [hunk 6]
```diff
@@ -13,16 +13,19 @@ This document defines the coding conventions and architecture boundaries for thi
...
 ## Core Principles
 
-### Code Quality
+We drive with a quality mindset in everything — planning, reviewing, and building.
+
 - **KISS** (Keep It Simple, Stupid) — Prefer simple solutions over clever ones
 - **DRY** (Don't Repeat Yourself) — Extract common patterns, but not prematurely
 - **YAGNI** (You Aren't Gonna Need It) — Don't add features until they're needed
-- **SRP** (Single Responsibility Principle) — Each module/function should do one thing
+- **SRP** (Single Responsibility Principle) — Each change, function or module should focusing on doing one thing
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply AGENTS.md:21]: nitpicking

File: .claude/hooks/enforce-allowlist.sh [hunk 7]
```diff
@@ -60,6 +77,39 @@ check_file_write() {
...
+  local i char next
+
+  for ((i = 0; i < ${#pattern}; i++)); do
+    char="${pattern:i:1}"
+
+    if [[ "$char" == "*" ]]; then
+      next="${pattern:i+1:1}"
+      if [[ "$next" == "*" ]]; then
+        regex+=".*"
+        i=$((i + 1))
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .claude/hooks/enforce-allowlist.sh]: Verdict: valid

     Current code in .claude/hooks/enforce-allowlist.sh:80 turns **/*.md into ^.*/[^/]*\.md$. That does not match README.md at repo root. I reproduced that directly.

     So cubic’s claim is correct: **/ is being treated as “one or more directory chars plus slash”, not “zero or more directories”. That is a real matcher bug.

     My opinion:
      - worth fixing
      - small, self-contained, improves Quest
      - add an integration test for root-level and nested matches

File: .skills/quest/agents/arbiter.md [hunk 8]
```diff
@@ -34,12 +34,45 @@ The Arbiter exists to **prevent spin** and enforce engineering pragmatism. It fi
 5. Produce a **synthesized verdict** with one of:
    - `iterate` — plan needs changes. Provide a focused, prioritized list of issues for the Planner.
    - `approve` — plan is good enough. Proceed to Builder.
-6. Write the verdict to `.quest/<id>/phase_01_plan/arbiter_verdict.md`
+6. Write the verdict to `.quest/<id>/phase_01_plan/arbiter_verdict.md.next`
 7. Synthesize canonical findings from both review markdown artifacts and write:
-   - `.quest/<id>/phase_01_plan/review_findings.json`
+   - `.quest/<id>/phase_01_plan/review_findings.json` (or the orchestrator-provided `.next` scratch path for atomic publish)
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .skills/quest/agents/arbiter.md]: Verdict: valid, but low leverage

     In .skills/quest/agents/arbiter.md:37, step 6 says verdict goes to .next, but step 7 still says review_findings.json with an “or maybe .next” caveat.

     That is inconsistent with the actual workflow/runtime contract, which is now scratch-first and atomic-publish later.

     My opinion:
      - technically correct comment
      - this is a docs/prompt cleanup, not a runtime bug
      - worth doing only because it is cheap and removes ambiguity for the arbiter prompt
      - not worth a standalone bigger change

File: scripts/quest_review_intelligence.py [hunk 9]
```diff
@@ -95,12 +97,33 @@ def _cmd_merge_findings(args: argparse.Namespace) -> int:
...
+    expected_phase = getattr(args, "expected_phase", None)
+    if expected_phase:
+        actual_phase = payload.get("phase") if isinstance(payload, dict) else None
+        if actual_phase != expected_phase:
+            errors = list(errors) + [
+                f"expected phase='{expected_phase}' but backlog has phase='{actual_phase}'"
+            ]
+    if getattr(args, "strict_plan_defaults", False):
+        errors = list(errors) + validate_plan_phase_defaults(payload)
+    item_count = len(payload.get("items", [])) if isinstance(payload, dict) else 0
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_review_intelligence.py]: Verdict: valid

     This one is real in current Quest. In scripts/quest_review_intelligence.py:122, item_count = len(payload.get("items", [])) will throw on {"items": null}.

     I reproduced it; it raises TypeError instead of returning structured validation errors.

     My opinion:
      - definite fix-now
      - small and correct
      - add a regression test for items: null and maybe non-list values

File: scripts/quest_runtime/artifacts.py [hunk 10]
```diff
@@ -17,9 +17,8 @@
     "arbiter": (
         "phase_01_plan",
         (
-            "arbiter_verdict.md",
-            "review_findings.json",
-            "review_backlog.json",
+            "arbiter_verdict.md.next",
+            "review_findings.json.next",
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_runtime/artifacts.py:21]: ### Invalid/ stale

Verdict: stale / not valid anymore

     Current Quest intentionally does this:
      - arbiter writes scratch artifacts: .next
      - orchestrator validates them
      - orchestrator builds review_backlog.json.next
      - only then does it atomically publish canonical files

     Current runtime reflects that on purpose:
      - scripts/quest_runtime/artifacts.py:17 expects only arbiter scratch artifacts plus handoff
      - tests/unit/test_artifacts.py:67 explicitly asserts arbiter uses .next artifacts and no backlog
      - workflow doc already describes build/validate/publish of canonical backlog after arbiter outpu

File: .skills/engineering-throughput-spreadsheet/SKILL.md [hunk 11]
```diff
@@ -45,144 +55,135 @@ If the user provides team names but not Jira source mappings, make conservative
...
-   - `GITHUB_REPO_FOR_RELEASE_TRACKING`
-5. Keep secrets out of logs and final answers.
+4. If repos were not explicitly provided by the user, resolve them from `.env`, show the exact list to the user, and require confirmation before collecting GitHub data.
+5. If the repo list is wrong, instruct the user to update the env variable(s) and restart the prompt after the env file is corrected.
+6. Keep secrets out of logs and final answers.
+7. Use the committed repo-confirmation script before the build:
+
+```bash
+python3 scripts/engineering_throughput_show_config.py \
+  --team-config <team-config.json> \
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .skills/engineering-throughput-spreadsheet/SKILL.md]: Good catch. cubic found a real doc mismatch here. I updated both committed CLI examples so `--team-config` is optional when the run uses only `--jira-csv-dir`; that landed in `f166c77`.
- Reviewed by Codex, in collaboration with KjellKod

File: engineering_throughput/date_ranges.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+        return f"{focus_year} {start_month}"
+    return f"{focus_year} {start_month}-{end_month}"
+
+
+def _is_month_aligned_start(value: date) -> bool:
+    return value.day == 1
+
+
+def _is_valid_focus_end(value: date, current_day: date) -> bool:
+    return value == current_day or value == _month_end(value)
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply engineering_throughput/date_ranges.py:50]: Good catch. cubic found a real validation gap here. I now reject future month-end `--date-end` values instead of treating them as valid partial periods, and added a regression test in `f166c77`.
- Reviewed by Codex, in collaboration with KjellKod

File: engineering_throughput/config.py [hunk 13]
```diff
@@ -0,0 +1,416 @@
...
+    return result
+
+
+def parse_iso_date(value: str | None) -> date | None:
+    """Parse an optional YYYY-MM-DD date string."""
+
+    if not value:
+        return None
+    try:
+        return date.fromisoformat(value)
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply engineering_throughput/config.py]: Good catch. cubic found a real config-validation bug here. `parse_iso_date` now rejects non-string JSON values with a validation error instead of leaking a `TypeError`, and there is a targeted regression test in `f166c77`.
- Reviewed by Codex, in collaboration with KjellKod

File: git_metrics/throughput_collect.py [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+                    continue
+                raise RuntimeError(messages)
+            return payload["data"]
+        raise RuntimeError("GraphQL request failed after retries")
+
+    def search_merged_prs(self, owner: str, repo: str, window: MonthWindow) -> list[dict[str, Any]]:
+        query_text = f"repo:{owner}/{repo} is:pr is:merged merged:{window.start.isoformat()}..{window.end.isoformat()}"
+        query = """
+        query($query: String!, $cursor: String) {
+          search(type: ISSUE, query: $query, first: 100, after: $cursor) {
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply git_metrics/throughput_collect.py:133]: Good catch. cubic found a real data-integrity risk here. The collector now checks GitHub search counts and recursively splits oversized date ranges before the 1,000-result GraphQL cap can truncate results; that landed in `f166c77` with regression coverage.
- Reviewed by Codex, in collaboration with KjellKod

File: tests/integration/test_engineering_throughput_build.py [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+import json
+import os
+import subprocess
+import sys
+from pathlib import Path
+
+import pytest
+
+
+FIXTURE_DIR = Path("tests/fixtures/engineering_throughput")
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply tests/integration/test_engineering_throughput_build.py:13]: Good catch. cubic found a real CI portability bug here. The missing throughput fixtures were added and unignored in `4566f4d`, and the integration test now runs on a clean checkout.
- Reviewed by Codex, in collaboration with KjellKod
